File: android/app/src/main/kotlin/org/gaindrive/android/ui/Load.kt
package org.gaindrive.android.ui

/**
 * The three states every screen has. A sealed type rather than a data class
 * with nullable fields, so "loaded but empty" and "not loaded yet" cannot be
 * confused — which is the bug that produces a flash of "nothing here" on every
 * screen open.
 */
sealed interface Load<out T> {
	data object Loading : Load<Nothing>
	data class Failed(val message: String) : Load<Nothing>
	data class Ready<T>(val value: T) : Load<T>
}

fun <T> Load<T>.valueOrNull(): T? = (this as? Load.Ready)?.value

File: ios/GainDrive/UI/Load.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// The three states every browse screen has, mirroring `ui/Load.kt`.
///
/// An enum rather than a struct with nullable fields, so **"loaded but empty"
/// and "not loaded yet" cannot be confused** — that confusion is the bug that
/// produces a flash of "nothing here" on every screen open.
enum Load<Value> {
	case loading
	case failed(String)
	case ready(Value)

	/// Android spells this `valueOrNull()`; a property is the Swift form of the
	/// same thing.
	var value: Value? {
		if case .ready(let value) = self { value } else { nil }
	}

	var isLoading: Bool {
		if case .loading = self { true } else { false }
	}
}

extension Load: Sendable where Value: Sendable {}
extension Load: Equatable where Value: Equatable {}
